solve validatateParameters issues

akshay-99 opened the pull request:
<!--
  Thank you for contributing! Please use this pull request (PR) template.


 In the description field of this PR, include "resolves #XXXX" tagging the issue you are fixing. If this PR addresses the issue but doesn't completely resolve it (ie the issue should remain open after your PR is merged), write "addresses #XXXX".-->
Resolves #4558 

Changes:
<!-- Add here what changes were made in this pull request and if possible provide links showcasing the changes. -->
1. Store argument types to increase speed a bit and to prevent
   repetitive logging
2. Create new JSON file parameterData.json from data.json and
   use it for the FES instead so as to reduce library size

Running on the following sketch for testing on Chrome
```js
let iterations = 10000000;
function setup() {
  console.time('timer random');
  for(let i = 0; i<iterations; i++){
    random();
  }
  console.timeEnd('timer random');

  console.time('timer randomGaussian');
  for(let i = 0; i<iterations; i++){
    randomGaussian(10, 2);
  }
  console.timeEnd('timer randomGaussian');

  console.time('timer color');
  for(let i = 0; i<iterations; i++){
    color(10, 10, 10);
  }
  console.timeEnd('timer color');
  
}
```

Before
```
timer random: 693.552978515625ms
timer randomGaussian: 1639.44287109375ms
timer color: 4085.77392578125ms
```

After
```
timer random: 581.656982421875ms
timer randomGaussian: 1254.043212890625ms
timer color: 2215.990234375ms
```

I found that the performance improvement is only significant for functions with a lot of formats and overloads like `color()`.
However, this now prevents repetitive logging.
Earlier
![image](https://user-images.githubusercontent.com/38867671/82273456-2ea88c80-999b-11ea-9bd2-902c08c3c78f.png)

Now
![image](https://user-images.githubusercontent.com/38867671/82273570-6c0d1a00-999b-11ea-90fe-be64121c7739.png)


And the size of the compiled library now comes down to 3.5 MB. The source on master gives a final js file of size 4.7 MB

data.json is built by yui docs and it allows us to preprocess the data in preprocessor.js. But since data.json is used for generating the reference pages, I am leaving it untouched and creating another file parameterData.json. I am doing this bit in preprocessor.js but I am not sure if it is the correct place to do this. So until that is confirmed, I am keeping this PR as draft.

EDIT:
`int()` doesn't use validateParameters, my bad. Removing it from the above sketch.
`randomGaussian` doesn't use validateParameters directly but makes calls to `random()` within it. So a speedup of random speeds up `randomGaussian` too.

#### PR Checklist
<!--
  To check any option, replace the "[ ]" with a "[x]". Be sure to check out how it looks in the Preview tab! Feel free to remove any portion of the template that is not relevant for your issue.
-->

- [x] `npm run lint` passes
- [ ] [Inline documentation] is included / updated
- [x] [Unit tests] are included / updated

[Inline documentation]: https://github.com/processing/p5.js/blob/master/contributor_docs/inline_documentation.md
[Unit tests]: https://github.com/processing/p5.js/tree/master/contributor_docs#unit-tests

stalgiag commented:
This is really exciting! There are some really big improvements here.

I believe that `preprocessor.js` is a good place to create the `parameterData.json`. What do you think @lmccart ?

@akshay-99 maybe it also makes sense to add tests for argument tree lookup to this PR? This can ensure that future changes keep these improvements in place and prevent unnecessary lookup and redundant logs.

akshay-99 commented:
> @akshay-99 maybe it also makes sense to add tests for argument tree lookup to this PR?

Yes that sounds good. I will add them soon.

akshay-99 commented:
On testing performance on this sketch
```js

let iterations = 10000000;
function setup() {

  console.time('timer');
  for(let i = 0; i<iterations; i++){
    random(10, 40);
  }
  console.timeEnd('timer');

  console.time('timer');
  for(let i = 0; i<iterations; i++){ 
    color([40, 40, 40]);
  }
  console.timeEnd('timer');

  console.time('timer');

  // drawing 10 million circles would take too long
  // hence iterations/10
  for(let i = 0; i<iterations/10; i++){ 
    circle(10, 10, 10);
  }
  console.timeEnd('timer');
}
```

The following was observed
Before:
```
timer: 1414.52099609375ms
timer: 4764.137939453125ms
timer: 3715.78515625ms
```
After: 
```
timer: 976.86376953125ms
timer: 2638.384033203125ms
timer: 3660.899169921875ms
```

Since `circle()` has only one format for the arguments, the time taken is pretty much the same as before.

In the latest commit before this, I have added more inline comments and changed some things for more simplicity and readability. @stalgiag I think this is now ready for review. Marking it as such.

lmccart commented:
this looks good to me, @stalgiag do you want to take a look over and merge?
agree, `preprocessor.js` makes sense as the location for this.

stalgiag commented:
Hi! Sorry I was away from the computer for a couple of days. Thanks for merging this @lmccart and congrats on this great start to your GSOC project @akshay-99 !

lmccart commented:
whoops that was an accidental merge, sorry meant to leave that one for you @stalgiag. anyway, hooray @akshay-99
